dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp

```diff
@@ -88,28 +88,38 @@ void FineGrainedShuffleWriter<ExchangeWriterPtr>::prepare(const Block & sample_b
     prepared = true;
 }
 
-template <class ExchangeWriterPtr>
-void FineGrainedShuffleWriter<ExchangeWriterPtr>::flush()
+template <class ExchangeWriterPtr, bool selective_block>
+void FineGrainedShuffleWriter<ExchangeWriterPtr, selective_block>::flush()
 {
     if (rows_in_blocks > 0)
         batchWriteFineGrainedShuffle();
 }
 
-template <class ExchangeWriterPtr>
-bool FineGrainedShuffleWriter<ExchangeWriterPtr>::isWritable() const
+template <class ExchangeWriterPtr, bool selective_block>
+bool FineGrainedShuffleWriter<ExchangeWriterPtr, selective_block>::isWritable() const
 {
     return writer->isWritable();
 }
 
-template <class ExchangeWriterPtr>
-void FineGrainedShuffleWriter<ExchangeWriterPtr>::write(const Block & block)
+template <class ExchangeWriterPtr, bool selective_block>
+void FineGrainedShuffleWriter<ExchangeWriterPtr, selective_block>::write(const Block & block)
 {
     RUNTIME_CHECK_MSG(prepared, "FineGrainedShuffleWriter should be prepared before writing.");
     RUNTIME_CHECK_MSG(
         block.columns() == dag_context.result_field_types.size(),
         "Output column size mismatch with field type size");
 
-    size_t rows = block.rows();
+    size_t rows = 0;
+    if constexpr (selective_block)
+    {
+        RUNTIME_CHECK(block.info.selective && !block.info.selective->empty());
```

Comment:
why need to throw error if selective->empty() is true? Seems this function does not assume input block is not empty

Reply:
The next PR will make sure every `block.info.selective` is filled with data.

Reply:
but I think just check `block.info.selective` is enough?

Reply:
If `block.info.selective->empty()` is true, it means it's an empty block.
The code path shouldn't reach here.